subprojects/plugins/src/main/java/org/gradle/jvm/component/internal/JvmSoftwareComponentInternal.java

```diff
@@ -0,0 +1,165 @@
+ * <p>
+ * Before publicizing this component, we also need to consider how component extensibility works.
+ * For example, for the java-library plugin we have the additional {@code api} and {@code compileOnlyApi}
+ * scopes. Do we expect that plugin to write a new component interface and extend this in order to add the
+ * proper getter methods? What about the concrete component class which implements that new interface? Does
+ * it extend the default implementation class? Is there a way we can avoid Java inheritance?
+ */
+public interface JvmSoftwareComponentInternal extends SoftwareComponent {
+
+    /**
+     * Configures this component to publish a javadoc jar alongside the primary artifacts. As a result,
+     * this method also configures the necessary configurations and tasks required to produce
+     * the javadoc artifact.
+     */
+    void withJavadocJar();
+
+    /**
+     * Configures this component to publish a sources jar alongside the primary artifacts. As a result,
+     * this method also configures the necessary configurations and tasks required to produce
+     * the sources artifact.
+     */
+    void withSourcesJar();
+
+    /**
+     * Get the {@link Jar} task which assembles the resources and compilation outputs into
+     * a single artifact.
+     *
+     * @return A provider which supplies the component's {@link Jar} task.
+     */
+    TaskProvider<Jar> getJar();
+
+    /**
+     * Get the {@link JavaCompile} task which compiles the Java source files into classes.
+     *
+     * @return A provider which supplies the component's {@link JavaCompile} task.
+     */
+    TaskProvider<JavaCompile> getCompileJava();
+
+    // TODO: This could use a better name. These "outputs" can be confused with a
+    // component's "real" outputs, which are its variants.
+    /**
+     * Get the resources and compilation outputs for this component which are used as
+     * inputs for the {@link #getJar() jar} task.
+     *
+     * @return The source set outputs.
+     */
+    SourceSetOutput getOutput();
+
+    /**
+     * Get this component's backing source set.
+     * <p>
+     * {@link SourceSet#getOutput()} and the classpath-returning methods on the returned
+     * source set should ideally be avoided in favor of the similarly-named methods on
+     * this component. The concept of source sets having a single set of outputs is only
+     * relevant for single-target components.
+     *
+     * @return This component's source set.
+     */
+    SourceSet getSources();
+
+    // TODO: For these below Configuration-returning methods, we should consider using a
+    // "Configuration" suffix (ex. getImplementationConfiguration) so that when we update
+    // the JVM ecosystem to migrate away from Configurations towards scopes/resolvers/variants,
+    // we can utilize the "good" names without breaking changes.
+
+    /**
+     * Gets the scope configuration for which to declare dependencies internal to the component.
```

Comment:
👎 to "scope configuration"

Reply:
Its really a shame you're not a fan of this terminology. IMO our "buckets" draw a lot of parallels to maven's scopes, and I think its a term that comes naturally and is something that many users would be comfortable with. IMO buckets are really a terrible name for this concept and before going public with an API for this, we need to think of something better. 

What are your major holdouts to calling these things scopes?

Reply:
It's just way too overloaded.  Also, I think when assembling a custom collection of related dependencies, the author will not be thinking in terms of "creating a scope" but more "populating a collection". 

For another thing, we're [already using the word "bucket" in our docs](https://docs.gradle.org/6.0.1/userguide/feature_variants.html): 

> Said differently, this allows you to declare the dependencies specific to a feature in their own "bucket", but everything is still compiled as a single source set.

Reply:
We also use `scope` in several places to refer to configurations which hold dependencies:
[Gradle calls the scope of a dependency a configuration](https://github.com/gradle/gradle/blob/master/subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/core_dependency_management.adoc?plain=1#L34-L36)
[Every dependency declared for a Gradle project applies to a specific scope](https://github.com/gradle/gradle/blob/master/subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/declaring_dependencies.adoc?plain=1#L9-L12)
[Similar to dependency declarations, dependency constraint declarations are scoped by configurations](https://github.com/gradle/gradle/blob/1101aa9196b00362761236e74a1fe68e4ad9cc67/subprojects/docs/src/docs/userguide/dep-man/03-controlling-transitive-dependencies/dependency_constraints.adoc?plain=1#L27)
[Gradle dependencies and constraints are declared and scoped to a configuration](https://github.com/gradle/gradle/blob/master/subprojects/docs/src/docs/userguide/jvm/java_platform_plugin.adoc?plain=1#L54)

Usages of `bucket`:
[It's just a bucket that can hold a list of dependencies](https://github.com/gradle/gradle/blob/master/subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/declaring_dependencies.adoc?plain=1#L85)
[declare the dependencies specific to a feature in their own "bucket"](https://github.com/gradle/gradle/blob/1101aa9196b00362761236e74a1fe68e4ad9cc67/subprojects/docs/src/docs/userguide/dep-man/04-modeling-features/feature_variants.adoc?plain=1#L103)
[Each dependency bucket returns a DependencyAdder](https://github.com/gradle/gradle/blob/1101aa9196b00362761236e74a1fe68e4ad9cc67/subprojects/docs/src/docs/userguide/jvm/jvm_test_suite_plugin.adoc?plain=1#L239)

Looking through all usages of these terms in the docs, I don't see one being use all that much more than the other, though seemingly scope is used a bit more often if I needed to pick a winner. The word scope is used twice explicitly in the "What are dependency configurations" section of the declaring_dependencies adoc page. 

Though, I don't think existing usage of either of these two terms should be used as a driver for choosing a new name. Seemingly, most of our documentation refers to these things as _configurations_ or _dependency configurations_ without any other term to describe them. 


> Also, I think when assembling a custom collection of related dependencies, the author will not be thinking in terms of "creating a scope" but more "populating a collection".

Can you elaborate on this? This would imply a name more like `DependencyCollection` than bucket

Reply:
> Can you elaborate on this? This would imply a name more like DependencyCollection than bucket

I thought part of the appeal of "bucket" is that it's a synonym for "bag", used in collection land to mean an unordered collection that (unlike a set) allows duplicates.

I think declaring a new configuration named say `database` or `logging` and adding all the related deps into it purely as a grouping mechanism, to be extended by perhaps multiple configurations that are ultimately resolved, allows you to do useful things.  You can apply constraints more granularly (than to your resolvable configuration like `implementation` as a whole) while documenting reasons for deps with `because`.  You can't duplicate this using version catalogs bundles (yet?).  It makes sense to me that the  different resolvable configurations are conceptually associated with something like a scope in the project's build process, but these extended configurations don't really seem like that to me, they're just different bunches of deps we want to group together.

We should definitely take a pass at making the language in those docs consistent after we decide on a term.  I was just surprised to see "bucket" there at all (and I must have read these particular pages before, I just didn't recall seeing the word).

Reply:
I don't want to hold up this PR while we make this decision. I've changed the wording for now to `dependency configuration`. 

> I thought part of the appeal of "bucket" is that it's a synonym for "bag", used in collection land to mean an unordered collection that (unlike a set) allows duplicates.

I was not aware of the context of why we chose the name bucket, which sort of makes sense, though doesn't dependency order matter? In any case why invent a new term at all? Is the intention to explicitly define a bucket as a "Bag of dependencies"? 

If that is the intended reasoning of the definition, I think we're missing an important point -- dependency configurations support a hierarchy and "bags" do not. A dependency configuration is more than a bag of dependencies. Its a bag of dependencies which order DOES matter (not a bag) and also supports extending other bags. I'm not saying scope is the right term here, but I'm don't think the bag/bucket term is the right path here. 

> It makes sense to me that the different resolvable configurations are conceptually associated with something like a scope in the project's build process, but these extended configurations don't really seem like that to me, they're just different bunches of deps we want to group together.

I'd be curious to hear more about these thoughts. While dependency configurations conceptually don't need to have actual semantic differences, in practice they do. In our first-party plugins at least, we don't often (or at all) create separate dependency configurations which do not have some particular semantic purpose. `api`, `implementation`, etc ARE semantically different, and I think there is some kind of conceptual "scope" for these dependency collections. 

But, I do agree that a dependency configuration doesn't necessarily _need_ to be used to define a "scope"

Reply:
I understand the resolved order will matter when a classpath is created, for all the usual reasons.  Will changing the order dependencies are added to a configuration influence the shape of the resolved graph in any way though?  I didn't think it would.

The fact that configurations are hierarchical is a good point to consider.  Should these be, at the end of all this?  Or should buckets merely provide a `include(otherBucket)` method or something (and consumables/resolvables specify what buckets they are using in another way as well) and these not be modeled with an extension relationship?

Reply:
I think the hierarchy does matter. To me, a strict non-hierarchical structure would mean an `include(otherBucket)` method would simply add all dependencies from one bucket to another. But what it really means is that the first bucket inherits all dependencies from the second. If you add new dependencies to the second after the `include` call, we'd expect the first to reflect those changes. 

In any case, I don't think we should be using the same `extendsFrom` terminology for the behavior of a consumable/resolvable "extending" from a bucket. 

In terms of dependency ordering affecting the graph, I am not sure. But the ordering is certainly significant.